src/main/java/com/allenai/ml/linalg/SparseVector.java

```diff
@@ -87,5 +89,5 @@ public double dotProduct(Vector other) {
           result += val * other.at(idx);
```

Comment:
Yeah, sparse on dense is fine with this kind of impl, and sparse on (primitive)map-backed sparse is also pretty ok (but the map-backed version has other perf deficiencies compared to the nice parallel array impl as well, so I totally approve of going with this one!).

The case where you are doing sparse on sparse, and they're both implemented under the hood as a (long[], double[]) pair, I've always just the same thing you do for efficient conjunction queries in an inverted index: just stagger walk the two index arrays until you exhaust one of them:

public double dot(long[] idx1, double[] val1, long[] idx2, double[] val2) {
    int offset1 = 0;
    int offset2 = 0;
    double result = 0;
    while (offset1 < idx1.length && offset2 < idx2.length) {
        if (idx1[offset1] < idx2[offset2]) {
            offset1++;
        } else if (idx1[offset1] > idx2[offset2]) {
            offset2++;
        } else {
            result += (val1[offset1] * val2[offset2]);
            offset1++;
            offset2++;
        }
    }
    return result;
}

This of course relies on having access to the backing arrays.